### PR TITLE
Add log streaming for event and job pages

### PR DESCRIPTION
**Pull request by willie-yao**
This PR aims to add live log streaming to the event and job pages.

### CODE REVIEW

File: v2/cli/term/event_page.go [hunk 1]
```diff
@@ -20,7 +20,10 @@ type eventPage struct {
 	eventInfo  *tview.TextView
 	workerInfo *tview.TextView
 	jobsTable  *tview.Table
+	eventLogs  *tview.TextView
+	logModal   tview.Primitive
 	usage      *tview.TextView
+	logsClient core.LogsClient
```
Comment (krancour) [inline v2/cli/term/event_page.go]: I wouldn't add a new field for this. We already have reliable access to this client via `e.apiClient.Events().Logs()`.

File: v2/cli/term/event_page.go [hunk 2]
```diff
@@ -35,14 +38,32 @@ func newEventPage(
 		eventInfo:  tview.NewTextView().SetDynamicColors(true),
 		workerInfo: tview.NewTextView().SetDynamicColors(true),
 		jobsTable:  tview.NewTable().SetSelectable(true, false),
+		eventLogs:  tview.NewTextView().SetDynamicColors(true),
 		usage: tview.NewTextView().SetDynamicColors(true).SetText(
-			"[yellow](F5) [white]Reload    [yellow](<-/Del) [white]Back    [yellow](ESC) [white]Home    [yellow](Q) [white]Quit", // nolint: lll
+			"[yellow](F5) [white]Reload    [yellow](<-/Del) [white]Back    [yellow](W) [white]Logs    [yellow](ESC) [white]Home    [yellow](Q) [white]Quit", // nolint: lll
```
Comment (krancour) [inline v2/cli/term/event_page.go]: It's not clear why "W" represents "Logs".

I would either change it to "L" for "Logs" or "W' for "Worker Logs".

File: v2/cli/term/event_page.go [hunk 3]
```diff
@@ -35,14 +38,32 @@ func newEventPage(
...
+		return tview.NewFlex().
+			AddItem(nil, 0, 1, false).
+			AddItem(tview.NewFlex().SetDirection(tview.FlexRow).
+				AddItem(nil, 0, 1, false).
+				AddItem(p, height, 1, false).
+				AddItem(nil, 0, 1, false), width, 1, false).
+			AddItem(nil, 0, 1, false)
+	}
+
+	e.logModal = modal(e.eventLogs, 85, 25)
```
Comment (krancour) [inline v2/cli/term/event_page.go]: I don't understand what the rationale might have been for defining a function here and then immediately using that function. Wouldn't this be simpler and more readable/understandable to just do this?

```
e.logModal = tview.NewFlex().
	AddItem(nil, 0, 1, false).
	AddItem(tview.NewFlex().
	SetDirection(tview.FlexRow).
	AddItem(nil, 0, 1, false).
	AddItem(e.eventLogs, 25, 1, false).
	AddItem(nil, 0, 1, false), 85, 1, false).
	AddItem(nil, 0, 1, false)
```

File: v2/cli/term/page_router.go [hunk 4]
```diff
@@ -63,13 +64,15 @@ func (r *pageRouter) loadEventPage(eventID string) {
 	r.loadPage(eventPageName, func() {
 		r.eventPage.refresh(eventID)
 	})
+	go r.eventPage.streamEventLog(eventID)
```
Comment (krancour) [inline v2/cli/term/page_router.go]: We definitely shouldn't start streaming these logs until the user actually elects to view them.

File: v2/cli/term/page_router.go [hunk 5]
```diff
@@ -63,13 +64,15 @@ func (r *pageRouter) loadEventPage(eventID string) {
...
 }
 
 // loadJobPage refreshes the job page and brings it into focus.
 func (r *pageRouter) loadJobPage(eventID, jobID string) {
 	r.loadPage(jobPageName, func() {
 		r.jobPage.refresh(eventID, jobID)
 	})
+	go r.jobPage.streamLog(eventID, jobID)
```
Comment (krancour) [inline v2/cli/term/page_router.go]: We definitely shouldn't start streaming these logs until the user actually elects to view them.

File: v2/cli/term/page_router.go (hunk 4 above)
Comment (willie-yao) [reply v2/cli/term/page_router.go]: I'm getting an issue where if I start the goroutine after the user hits "L", it doesn't show logs until I pull up the logs modal again. Is it okay to load logs upon loading the event page, or is this a separate issue?

Comment (willie-yao) [reply v2/cli/term/page_router.go]: Actually it's even weirder, I can type any key and it would refresh and print out all the correct logs. I wonder what could be causing this

Comment (krancour) [reply v2/cli/term/page_router.go]: Don't worry about log streaming yet. Just use static placeholder text for now.

File: v2/cli/term/event_page.go [hunk 6]
```diff
@@ -113,7 +108,9 @@ func (e *eventPage) refresh(eventID string) {
 			switch evt.Rune() {
 			case 'r', 'R': // Reload
 				e.router.loadEventPage(eventID)
-			case 'w', 'W':
+			case 'l', 'L':
+				e.eventLogs.SetText("Placeholder logs")
+				go e.streamEventLog(eventID)
```
Comment (krancour) [inline v2/cli/term/event_page.go]: This is problematic because you're launching a goroutine and giving yourself no way to end that goroutine. I'd really hold off on trying to stream at this point. If you get the UI elements working, we can pair on how to do the streaming cleanly.

Comment (willie-yao) [reply v2/cli/term/event_page.go]: Ok sounds good, I'll remove these calls for now

File: v2/cli/term/event_page.go [hunk 7]
```diff
@@ -35,14 +37,28 @@ func newEventPage(
...
 	e.eventInfo.SetBorder(true).SetBorderColor(tcell.ColorYellow)
 	e.workerInfo.SetBorder(true).SetBorderColor(tcell.ColorYellow).
 		SetTitle("Worker")
 	e.jobsTable.SetBorder(true).SetTitle("Jobs")
+	e.eventLogs.SetBorder(true).SetTitle("Logs (<-/Del) Quit")
+
+	// Returns a new primitive which puts the provided primitive in the center and
+	// sets its size to the given width and height.
+	e.logModal = tview.NewFlex().
+		AddItem(nil, 0, 1, false).
```
Comment (krancour) [inline v2/cli/term/event_page.go]: For my own edification-- are the several `AddItem(nil, 0, 1, false)` all about putting padding the layout with dead space so the page appears as a floating window? This is fine. I just want to make sure I understand correctly. I am no expert with this UI toolkit.

File: v2/cli/term/event_page.go [hunk 8]
```diff
@@ -20,6 +20,8 @@ type eventPage struct {
 	eventInfo  *tview.TextView
 	workerInfo *tview.TextView
 	jobsTable  *tview.Table
+	eventLogs  *tview.TextView
```
Comment (krancour) [inline v2/cli/term/event_page.go]: I would suggest changing this to `workerLogs`.

Events are really just "things that happened." Workers are the things that _handle_ the events. So events don't technically have a status. They don't technically have logs. They _do_ each have an associated worker and its worker has a status, logs, etc.

File: v2/cli/term/event_page.go [hunk 9]
```diff
@@ -35,14 +37,28 @@ func newEventPage(
...
+	// sets its size to the given width and height.
+	e.logModal = tview.NewFlex().
+		AddItem(nil, 0, 1, false).
+		AddItem(tview.NewFlex().
+			SetDirection(tview.FlexRow).
+			AddItem(nil, 0, 1, false).
+			AddItem(e.eventLogs, 25, 1, false).
+			AddItem(nil, 0, 1, false), 85, 1, false).
+		AddItem(nil, 0, 1, false)
+
```
Comment (krancour) [inline v2/cli/term/event_page.go]: A lot of this smells reusable to me.

I think we should extract this stuff into its own component that can be used across the event page (for displaying worker logs) and the job page (for displaying job logs).

I know that as things stand right now, the job page shows logs in the "table area" below the job details. This _probably_ isn't the right way to go.

Jobs can be composed of multiple containers. We should probably use that table area to list the job's containers. Selecting a container can pop up the logs for that container-- which leads me back to the idea that what you've done here on the event page is something we should consider making reusable for both the event and job page.

In fact...

Let's put a pin in this PR for now and have you focus on a new PR that lists a job's containers in the table area of the job page (selecting one of them doesn't need to do anything for now). Once we get that in, then we can revistiv this PR and figure out how best to make this little log pop-up reusable.

**Comment by krancour** [comment]
Left some comments inline.

I ❤️  how this _looks_. We have some work to do to make this log display component reusable across two different pages, but you're definitely on the right track in terms of what this should look like.

File: v2/cli/term/event_page.go (hunk 7 above)
Comment (willie-yao) [reply v2/cli/term/event_page.go]: Yes, this is to make the textfield appear to be a floating window

File: v2/cli/term/page_router.go [hunk 10]
```diff
@@ -102,6 +115,34 @@ func (r *pageRouter) loadPage(pageName string, fn func()) {
...
+			select {
+			case <-ticker.C:
+				fn()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
```
Comment (krancour) [inline v2/cli/term/page_router.go]: Is the only difference between this and the existing `loadPage()` function that one hides the other pages (`r.SwitchToPage()`) and the other doesn't (`r.ShowPage()`)?

If that is the case, we should figure out how to combine those into one function so that future updates don't have to be applied in two place.

We could just add something like a `hideOtherPages bool` arg to the argument list or something.

File: v2/cli/term/event_page.go [hunk 11]
```diff
@@ -20,6 +20,8 @@ type eventPage struct {
 	eventInfo  *tview.TextView
 	workerInfo *tview.TextView
 	jobsTable  *tview.Table
+	workerLogs *tview.TextView
+	logModal   tview.Primitive
```
Comment (krancour) [inline v2/cli/term/event_page.go]: Should these, and related changes, be undone?

File: v2/cli/term/page_router.go (hunk 10 above)
Comment (willie-yao) [reply v2/cli/term/page_router.go]: I was thinking of doing that, since there is a lot of repeated code. I created a whole new function just to test it out and forgot to run it by you. Gonna fix now!

File: v2/cli/term/page_router.go [hunk 12]
```diff
@@ -86,9 +100,14 @@ func (r *pageRouter) loadPage(pageName string, fn func()) {
...
 	var ctx context.Context
 	ctx, r.cancelRefreshFn = context.WithCancel(context.Background())
-	r.SwitchToPage(pageName) // Bring the page into focus
-	fn()                     // Synchronously refresh the page once
-	go func() {              // Start auto-refreshing
+	if hideOtherPages {
+		r.SwitchToPage(pageName) // Focus page and hide background
+	} else {
+		r.ShowPage(pageName) // Focus page and keep background
+		r.app.SetFocus(r.logPage.logText)
```
Comment (willie-yao) [inline v2/cli/term/page_router.go]: @krancour After consolidating the `showPage` function, I realized that I needed to set the focus to the `logText` page. I feel like it is a bit off to pass in a `textField` as a parameter since non-floating windows wouldn't need to, so it is hard-coded in for now. This may cause problems later down the line for pop-up windows that display something other than a `logText`. Do you think it would be okay to pass in a `textField` to `loadPage` as well?

File: v2/cli/term/event_page.go [hunk 13]
```diff
@@ -36,12 +36,13 @@ func newEventPage(
...
 		usage: tview.NewTextView().SetDynamicColors(true).SetText(
-			"[yellow](F5) [white]Reload    [yellow](<-/Del) [white]Back    [yellow](ESC) [white]Home    [yellow](Q) [white]Quit", // nolint: lll
+			"[yellow](F5) [white]Reload    [yellow](<-/Del) [white]Back    [yellow](L) [white]Logs    [yellow](ESC) [white]Home    [yellow](Q) [white]Quit", // nolint: lll
 		),
 	}
 	e.eventInfo.SetBorder(true).SetBorderColor(tcell.ColorWhite)
-	e.workerInfo.SetBorder(true).SetTitle(" Worker ")
-	e.jobsTable.SetBorder(true).SetTitle(" Jobs ")
+	e.workerInfo.SetBorder(true).SetTitle("Worker")
+	e.jobsTable.SetBorder(true).SetTitle("Jobs")
```
Comment (krancour) [inline v2/cli/term/event_page.go]: The extra padding was deliberate.

File: v2/cli/term/page_router.go (hunk 12 above)
Comment (krancour) [reply v2/cli/term/page_router.go]: Is there a reason that `r.app.SetFocus(pageName)` wouldn't work here? Or am I misunderstanding you?

Comment (willie-yao) [reply v2/cli/term/page_router.go]: `SetFocs()` takes in a `tview.Application` and not a string, unlike `ShowPage`

File: v2/cli/term/event_page.go (hunk 13 above)
Comment (willie-yao) [reply v2/cli/term/event_page.go]: Which padding did you mean? I removed the `logModal` since that is defined inside `log_page.go` now

File: v2/cli/term/page_router.go (hunk 12 above)
Comment (krancour) [reply v2/cli/term/page_router.go]: I see. I'll take a look in the AM.

File: v2/cli/term/event_page.go (hunk 13 above)
Comment (krancour) [reply v2/cli/term/event_page.go]: You removed extra spaces I placed around the titles.

Comment (willie-yao) [reply v2/cli/term/event_page.go]: Oops, must have missed that during the rebase. Fixing that now

File: v2/cli/term/page_router.go [hunk 14]
```diff
@@ -49,46 +49,47 @@ func NewPageRouter(
...
 // focus.
 func (r *pageRouter) loadLogPage(page *page, eventID, jobID string) {
-	// r.loadPage(logPageName, func() {
-	// 	r.logPage.refresh(*page, eventID, jobID)
-	// })
-	r.loadPage(logPageName, false, func() {
+	r.loadPage(logPageName, func() {
 		r.logPage.refresh(*page, eventID, jobID)
-	})
+	}, r.logPage.logText)
```
Comment (willie-yao) [inline v2/cli/term/page_router.go]: @krancour I modified `loadPage` to take in a page to focus on as a variadic input since there are no conditional inputs in Go. Would this make more sense than passing in a parameter strictly for a focus page, and having non-focus pages pass in a `nil` value?

Comment (krancour) [reply v2/cli/term/page_router.go]: See https://github.com/willie-yao/brigade/pull/1 for a cleaner alternative.

Comment (willie-yao) [reply v2/cli/term/page_router.go]: LGTM, thanks for the suggestion

File: v2/cli/term/log_page.go [hunk 15]
```diff
@@ -28,8 +31,11 @@ func newLogPage(
 		logText: tview.NewTextView().SetDynamicColors(true),
 	}
 
+	// GitHub places a restriction on the text field for a Check Run at 65535
+	// characters, so we use this as a maximum and truncate if needed.
```
Comment (krancour) [inline v2/cli/term/log_page.go]: This comment isn't relevant here.

File: v2/cli/term/log_page.go [hunk 16]
```diff
@@ -28,8 +31,11 @@ func newLogPage(
...
+	l.maxBytes = 65535
 	l.logText.SetBorder(true).SetTitle("Logs (<-/Del) Quit")
-	l.logString = "Waiting for logs..."
+	l.logBuf, _ = circbuf.NewBuffer(l.maxBytes)
```
Comment (krancour) [inline v2/cli/term/log_page.go]: Don't ignore errors...

Comment (willie-yao) [reply v2/cli/term/log_page.go]: How should I deal with errors inside of `newLogPage()`? Should I just return an empty page with an error message, since the buffer couldn't be initialized?

**Comment by krancour** [comment]
I built from this branch and for both workers and jobs, I get nil pointer dereferences almost every time I try to view logs.

**Comment by willie-yao** [comment]
That's weird, I never get a nil pointer dereference. Let me try again and see how it goes. Are you on the dogfood cluster? @krancour

**Comment by willie-yao** [comment]
Looks like I didn't update this branch from after I fixed the event color issue that caused a nil pointer dereference. I'm updating that now and seeing if I get any more issues. This wasn't an issue when I tested last week so I'm not sure what happened

File: v2/cli/term/log_page.go (hunk 16 above)
Comment (krancour) [reply v2/cli/term/log_page.go]: I think there may be two reasonable ways to deal with errors.

Yes... one of them would be development of an error page, but I'd about the complexity found therein creating the possibility that an initial error is obscured by an error in the error page itself.

You know tview better than I do... is there a way to just shut the UI down on command and then display an error to stderr like normal?

Comment (krancour) [reply v2/cli/term/log_page.go]: I should add that centralizing the error handling for term probably needs to be its own separate PR.

To account for my "don't ignore errors" comment, for now, just add a comment with a TODO.

Comment (willie-yao) [reply v2/cli/term/log_page.go]: Sure. I made a change to initialize the buffer inside the log page's load function instead, so if there is an error, I can display that to the floating log box that comes up and return before starting the two goroutines for streaming logs. 

This would make displaying the error message possible, since I can create the buffer after creating the textview. I have a feeling that this would be a better user experience than shutting down the entire UI. What do you think?

Comment (krancour) [reply v2/cli/term/log_page.go]: It works well for this case. It doesn't work well for the general case. It's fine for now, but longer term, we need something else.

File: v2/cli/term/event_page.go [hunk 17]
```diff
@@ -84,6 +90,11 @@ func (e *eventPage) refresh(eventID string) {
 			switch evt.Rune() {
 			case 'r', 'R': // Reload
 				e.router.loadEventPage(eventID)
+			case 'l', 'L':
+				// e.workerLogs.SetText("Placeholder logs")
+				// e.router.ShowPage("Event Logs")
+				// e.router.app.SetFocus(e.workerLogs)
```
Comment (krancour) [inline v2/cli/term/event_page.go]: You probably meant to remove these lines.

File: v2/cli/term/log_page.go (hunk 16 above)
Comment (willie-yao) [reply v2/cli/term/log_page.go]: Yeah I was thinking the same. We can add a standardized system for error handling in the future.

File: v2/cli/term/page_router.go [hunk 18]
```diff
@@ -40,41 +41,92 @@ func NewPageRouter(
...
+			r.jobPage.refresh(ctx, eventID, jobID)
+		},
+		true,
+	)
+}
+
+// loadLogPage loads a floating window that displays logs and brings it into
+// focus.
+func (r *pageRouter) loadLogPage(page *page, eventID, jobID string) {
+	// go r.logPage.streamLogs(eventID, jobID, quit)
```
Comment (krancour) [inline v2/cli/term/page_router.go]: I think you meant to remove this line.

File: v2/cli/term/page_router.go [hunk 19]
```diff
@@ -40,41 +41,92 @@ func NewPageRouter(
...
+		func(ctx context.Context) {
+			r.jobPage.refresh(ctx, eventID, jobID)
+		},
+		true,
+	)
+}
+
+// loadLogPage loads a floating window that displays logs and brings it into
+// focus.
+func (r *pageRouter) loadLogPage(page *page, eventID, jobID string) {
```
Comment (krancour) [inline v2/cli/term/page_router.go]: Do we really need this `page` arg? What are we doing with it?

Comment (willie-yao) [reply v2/cli/term/page_router.go]: Oops, this was left over from when I passed in a page to bring into focus, rather than calling `loadEventPage()` and `loadJobPage()`. Good catch, I'm surprised VSCode didn't catch an unused parameter
